[Mempool] Improve peer priority selection.

JoshLind opened the pull request:
Note: most of this PR is new unit tests.

## Description
This PR improves the peer selection logic for mempool (i.e. deciding which peers to forward transactions to):
- Currently, the logic just selects peers based on network types and roles, but this is somewhat inefficient (especially when nodes define seeds that override more peformant peers).
- To avoid this, the PR updates the selection logic to prioritize peers based on: (i) network type; (ii) distance from the validators (e.g., to avoid misconfigured and/or disconnected VFNs); and (iii) peer ping latencies (i.e., to favour closer peers).
- To avoid excessively reprioritizing peers (which can be detrimental to mempool under load), we only update the peer priorities when: (i) our peers change; (ii) we're still waiting for the peer monitoring service to populate the peer ping latencies; or (iii) if neither (i) or (ii) is true, we update every ~10 minutes (configurable). This avoids overly reprioritizing peers at steady state.

## Testing Plan
New and existing test infrastructure. I also ran several PFN-only tests to ensure that the average broadcast latencies are reduced.

sitalkedia commented on mempool/src/shared_mempool/priority.rs:
```diff
@@ -0,0 +1,728 @@
+    pub fn ready_for_update(&self, peers_changed: bool) -> bool {
+        // If our peers have changed, or we haven't observed ping latencies
+        // for all peers yet, we should update the prioritized peers again.
+        if peers_changed || !self.observed_all_ping_latencies {
```
Can you explain the reasoning behind updating this if we have not observed ping latency for all peers?

sitalkedia replied:
I see, makes sense. It might be better to provide this context with comment in the code.

sitalkedia commented on mempool/src/shared_mempool/priority.rs:
```diff
@@ -0,0 +1,728 @@
+            .iter()
+            .sorted_by(|peer_a, peer_b| {
+                let ordering = &self.peer_comparator.compare(peer_a, peer_b);
+                ordering.reverse() // Prioritize higher values (i.e., sorted by descending order)
```
@JoshLind - May be I am missing something but this seems to be returning the list in ascending order of priority - because the `peer_comparator` is returning the highest priority first?

JoshLind replied:
It's because `sorted_by` sorts in [ascending order](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.sorted_by), but we want high priority peers to be at the beginning of the list (i.e., descending order, as mempool selects peers from the front of the list). So, the easiest way to do that is reverse the ordering here 😄

JoshLind replied:
(We can also do this by flipping peer_a and peer_b, if that's clearer 🤔)

sitalkedia replied:
Thanks - this makes sense to me.

sitalkedia reviewed (approved):
LGTM

bchocho commented on mempool/src/shared_mempool/priority.rs:
```diff
@@ -0,0 +1,728 @@
+        // Update the prioritized peers
+        let mut prioritized_peers = self.prioritized_peers.write();
+        if new_prioritized_peers != *prioritized_peers {
+            info!(
```
Can we add an explicit metric for how many prioritized peers were changed? This can help us debug in case there are issues with changing frequently.

bchocho reviewed (commented):
Can we gate the priority sorting based on ping times?

One case I am concerned about is when there are a lot of outstanding items in mempool, getting a new peer in the priority peers means all the outstanding items have to be broadcasted to the new peer (in FIFO order), incurring significant latency for incoming transactions as they wait in a queue to be broadcasted. Not sure how problematic this could be in practice though.

There's no foolproof solution I can think about to resolve this. Here's some ideas--
* Constrain the number of peers to be re-prioritized in each 10 min interval to half of the peers (1 in our case)?
* For a new peer, constrain the number of "old" txns we forward to the new peer.

JoshLind commented:
> Can we gate the priority sorting based on ping times?

As discussed offline, I've updated the PR to do this. 😄 The "intelligent" sorting is enabled by default, but it can be disabled in the node config. If it is disabled, it means mempool will retain the same behaviour as before (i.e., only reprioritize on peer changes, and use a simpler prioritization algorithm). 

Note, however, that I have modified the original behaviour slightly: instead of using the "peer role" in the simple prioritization algorithm, we no longer use it, and instead sort by network and then peer hash. This is because the peer roles are imprecise, so it makes sense to fallback to random selection (instead of prioritizing seeds), when the feature is disabled.

Let me know what you think!

bchocho reviewed (approved):
Thanks! One comment on the metrics, but approving to unblock

JoshLind commented on mempool/src/counters.rs:
```diff
@@ -429,6 +430,20 @@ pub fn shared_mempool_pending_broadcasts(peer: &PeerNetworkId) -> IntGauge {
 }
 
+/// Counter tracking the number of peers that changed priority in shared mempool
+static SHARED_MEMPOOL_PRIORITY_CHANGE_COUNT: Lazy<IntGaugeVec> = Lazy::new(|| {
```
Aah, thanks @bchocho -- I missed `IntGauge`  🤦  Will change to that. Note: I'm not sure `Histogram` makes sense because we update this metric really infrequently. It's probably easiest to just see the raw gauge value instead of try to work with averages, rate changes, etc.
